### DISCUSSION TITLE
Adding fulfillment types to orders

### DISCUSSION THREAD

**Discussion by Mrkbingham** [NikolaGavric94/laravel-square#87] [Ideas]
I'm interested in adding in fulfillment types to this repo.  Curious if that would be something you'd like to integrate, or a notable enough departure that I should create my own fork?  Happy to provide a high-level overview of my plan if that's of any interest

https://github.com/square/square-php-sdk/blob/master/doc/models/order-fulfillment.md

**Comment by NikolaGavric94** [comment]
Hello Mark,

I would be interested to hear more about this, so please do share more
details

**Reply by Mrkbingham** [reply]
Great!  I've done a little bit of work, and this is still definitely in-progress with a few todos noted.  I'd like to hold off on adding in support for distinct line-item delivery methods as that adds a lot of complexity and seems like a real edge-case.  Otherwise, all to-dos in the code would be completed.

In summary, the outline would be:
1. Create migrations and models that all represent distinct fulfillment types (Delivery, Pickup, Shipment)
2. Add a single FulfillmentBuilder that contains all the logic for the three fulfillment types.
3. Add the fulfillment builder to the OrderBuilder and SquareRequestBuilder.
4. Add add-er and get-er's to the SquareServiceContract facade.
5. Create factories/tests for all scenarios.

I would prefer to add in all three fulfillment types at once, so the development flow would go something like (assuming each has tests added along the way):
1. Add the migrations
2. Add the fulfillment builders for models
3. Add the fulfillment builders from arrays

Alternatively, I could add in each fulfillment type one-at-a-time, but the disadvantage there is that this could require additional migrations for any mis-steps when determining the relationships between the individual fulfillment type models.  That flow would probably look like this (assuming we pick adding in pickup fulfillment types first):
1. Add a the fulfillment and pickup migrations
2. Add the fulfillment builder and pickup builders (both model and arrays)


In summary, the goal would be to add models, builders and related pivot-models for any relationships required to support this.  Here's a link to compare the current progress:
https://github.com/NikolaGavric94/laravel-square/compare/master...Mrkbingham:laravel-square:add_fulfillments?expand=1

Very open to any feedback on this!

**Comment by NikolaGavric94** [comment]
Hey Mark,

Thanks for coming back to me and with a lot of code changes done already,
if you could please create a PR with suggested changes and in the meantime
I will research this topic further and come back to you once I'm done with
initial research, One thing why this feature might be problematic to
implement could be because of a lot of use cases and limitations might
apply to fulfillments in general. If you could finalize the code changes in
the meantime and test coverage for all the possible scenarios and edge
cases that would be awesome.

Thanks

**Reply by Mrkbingham** [reply]
Sounds good!  I've got some real-world use-cases I can run through with existing Square connections to see what edge-cases I might be missing here.  I've wrapped up the fulfillment array and model-based tests, but still need to add in support for a recipient tied to the specific fulfillment.  I'll put together a PR here in the next week or so! Once that's done, I'll start running through some of my square order-syncing to make sure the real-world data supports the mechanism for syncing/creation I've assumed in the tables and models.

**Reply by Mrkbingham** [reply]
Okay, quick follow-up here @NikolaGavric94 .  I've created a new PR here: https://github.com/NikolaGavric94/laravel-square/pull/88

Happy to discuss any of these changes, I believe I've covered all the base scenarios and will start testing integrating this into my app to ensure the assumption about the square data models is accurate.  If there's anything missing, especially any tests you think I might've overlooked, I'm happy to update it!

**Reply by Mrkbingham** [reply]
Also, I know I'm hitting you with a lot of pings here - is there any chance you'd be able to add me as able to create a branch on your repo?  I'd like to be able to try integrating the changes I've made to my other application by using your existing composer package and targeting a non-master branch.

**Reply by Mrkbingham** [reply]
Nvm, I wasn't aware I could just use my own fork, I was under the impression I'd need to create an entirely new composer project.  Since I'm able to continue testing away and developing in my app, I don't feel as much of a sense of urgency towards integrating this change into the core-repository here.  But I'm still more than happy to work towards integrating this in to the core-repo of course!

cf. https://medium.com/swlh/using-your-own-forks-with-composer-699358db05d9
